src/qibolab/instruments/qm.py

```diff
@@ -623,21 +635,25 @@ def play(self, qubits, sequence, nshots, relaxation_time):
             for qmpulse in qmsequence.ro_pulses:
                 threshold = qubits[qmpulse.pulse.qubit].threshold
                 iq_angle = qubits[qmpulse.pulse.qubit].iq_angle
-                qmpulse.declare_output(threshold, iq_angle)
+                qmpulse.declare_output(threshold, iq_angle, raw_adc)
 
             with for_(n, 0, n < nshots, n + 1):
-                self.play_pulses(qmsequence, relaxation_time)
+                self.play_pulses(qmsequence, relaxation_time, raw_adc)
 
             with stream_processing():
                 # I_st.average().save("I")
                 # Q_st.average().save("Q")
                 # n_st.buffer().save_all("n")
                 for qmpulse in qmsequence.ro_pulses:
                     serial = qmpulse.pulse.serial
-                    qmpulse.I_st.buffer(nshots).save(f"{serial}_I")
-                    qmpulse.Q_st.buffer(nshots).save(f"{serial}_Q")
-                    if qmpulse.threshold is not None:
-                        qmpulse.shots.buffer(nshots).save(f"{serial}_shots")
+                    if raw_adc:
+                        qmpulse.raw_adc.input1().average().save(f"{serial}_I")
+                        qmpulse.raw_adc.input2().average().save(f"{serial}_Q")
+                    else:
+                        qmpulse.I_st.buffer(nshots).save(f"{serial}_I")
+                        qmpulse.Q_st.buffer(nshots).save(f"{serial}_Q")
+                        if qmpulse.threshold is not None:
+                            qmpulse.shots.buffer(nshots).save(f"{serial}_shots")
```

Comment:
Even the content of this `for` loop is a nice candidate for a separate function: it only depends on `qmpulse`, `raw_adc`, and `nshots`.

I'd leave it up to you to find a good name for it :)